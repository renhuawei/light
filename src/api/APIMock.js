import fetchMock from 'fetch-mock';
import pathToRegexp from 'path-to-regexp';
import { getRandomInt } from 'utils';

import providers from './mocks/providers.json';

const delay = () => new Promise(res => setTimeout(res, 500));

export function mockAPI() {
  fetchMock
    .get(`/api/providers`, () => delay().then(() => providers))
    .get('express:/api/providers/:id', url => {
      const match = pathToRegexp('/api/providers/:id').exec(url);
      return delay().then(() => {
        const provider = providers.data.find(({ id }) => id == match[1]);
        return { success: true, data: provider };
      });
    })
    .post('/api/refilAccount', () => {
      return delay().then(() => {
        if (getRandomInt(0, 2) === 1) {
          throw {
            message: 'Sorry! Service is temporary unavailable.'
          };
        }
        return { success: true };
      });
    });

  fetchMock.spy();
}

export default mockAPI;
