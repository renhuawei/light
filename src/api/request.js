import 'whatwg-fetch';

async function request(url, config = {}) {
  const headers = new Headers();
  headers.append('Accept', 'application/json');

  for (const newHeader in config.headers) {
    headers.set(newHeader, config.headers[newHeader]);
  }

  if (!(config.body instanceof FormData)) {
    headers.append('Content-Type', 'application/json');
  }

  if (config.json) {
    config.body = JSON.stringify(config.json);
  }

  const finalConfig = {
    method: 'get',
    ...config,
    headers
  };

  try {
    const response = await fetch(url, finalConfig);
    const contentType = response.headers.get('content-type');
    const isSerialisable =
      contentType && contentType.includes('application/json');

    if (response.status === 200 || response.status === 201) {
      if (isSerialisable) {
        return response.json();
      }
      return response.text();
    } else if (response.status >= 400) {
      return Promise.reject({
        failure: true,
        errors: [{ message: response.statusText }]
      });
    }
  } catch (err) {
    return Promise.reject({
      failure: true,
      errors: [{ message: err.message }]
    });
  }
}

export default request;
