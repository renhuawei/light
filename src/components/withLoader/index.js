import React from 'react';
import Loader from 'components/Loader';

function withLoader(WrappedComponent, request) {
  return class PageContainer extends React.Component {
    state = { status: 'loading' };

    async componentDidMount() {
      const { data } = await request(this.props.match);
      this.setState({ status: 'loaded', data });
    }

    render() {
      if (this.state.status === 'loading') {
        return <Loader>Loading...</Loader>;
      }

      return <WrappedComponent {...this.props} data={this.state.data} />;
    }
  };
}

export default withLoader;
