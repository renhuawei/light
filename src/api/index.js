import request from './request';
import mockAPI from './APIMock';

if (process.env.API_MODE === 'MOCK' || process.env.NODE_ENV === 'test') {
  mockAPI();
}

export const getProviders = async () => await request('/api/providers');

export const getProvider = async id => await request(`/api/providers/${id}`);

export const refillAccount = async payload =>
  await request(`/api/refilAccount`, {
    method: 'post',
    json: payload
  });
