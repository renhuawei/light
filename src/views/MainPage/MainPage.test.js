import React from 'react';
import { shallow } from 'enzyme';

import { MainPage } from './index';
import withLoader from 'components/withLoader';
import { getProviders } from 'api';

describe('MainPage', () => {
  it('render correctly', async () => {
    const promise = getProviders();
    const loadData = () => promise;
    const MainPageWrapper = withLoader(MainPage, loadData);
    const component = shallow(<MainPageWrapper />);

    //Render initial loading state
    expect(component).toMatchSnapshot();

    //Render MainPage Component after request is resolved
    await promise;
    expect(component).toMatchSnapshot();
  });
});
