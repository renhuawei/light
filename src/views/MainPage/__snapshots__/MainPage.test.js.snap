// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MainPage render correctly 1`] = `
<styled.div>
  Loading...
</styled.div>
`;

exports[`MainPage render correctly 2`] = `
<MainPage
  data={
    Array [
      Object {
        "id": 1,
        "name": "MTS",
      },
      Object {
        "id": 2,
        "name": "Beeline",
      },
      Object {
        "id": 3,
        "name": "Tele2",
      },
      Object {
        "id": 4,
        "name": "Megafon",
      },
    ]
  }
/>
`;
